File: Frontend/src/lib/services/tasks.ts
import { CacheTasks, Task } from "@/interfaces/entities/task";
import { Result } from "@/types/result";
import { invoke } from "@tauri-apps/api/core";

// ======================
// DB
// ======================

async function selectTask(env: string, task?: Task): Promise<Result<Task[]>> {
    try {
        const data = await invoke<Task[]>("select_task", {
            env,
            query: task,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function selectTasks(env: string, task?: Task): Promise<Result<Task[]>> {
    try {
        const data = await invoke<Task[]>("select_tasks", {
            env,
            query: task,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function updateTask(env: string, task: Task): Promise<Result<Task>> {
    try {
        const data = await invoke<Task>("update_task", { env, task });
        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

// ======================
// CACHE
// ======================

async function getActiveTasks(env: string): Promise<Result<CacheTasks>> {
    try {
        const data = await invoke<CacheTasks>("select_active_tasks", {
            env,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function startActiveTasks(env: string): Promise<Result<void>> {
    try {
        await invoke("start_active_tasks", { env });
        return { ok: true, data: undefined };
    } catch (error) {
        return { ok: false, error };
    }
}

async function stopActiveTasks(env: string): Promise<Result<void>> {
    try {
        await invoke("stop_active_tasks", { env });
        return { ok: true, data: undefined };
    } catch (error) {
        return { ok: false, error };
    }
}

// ======================
// EXPORT SERVICE
// ======================

export const taskService = {
    // DB
    select: selectTask,
    selectAll: selectTasks,
    update: updateTask,

    // Cache
    getActive: getActiveTasks,
    startActive: startActiveTasks,
    stopActive: stopActiveTasks,
};

File: Frontend/src-tauri/src/command/tasks.rs
use crate::core::{
    select_active_tasks_core, select_task_core, select_tasks_core, start_active_tasks_core,
    stop_active_tasks_core, update_task_core,
};
use crate::models::entities::tasks::Model;
use crate::models::structs::{CacheTasks, TaskRequest};

// db
#[tauri::command]
pub async fn select_task(env: String, query: Option<TaskRequest>) -> Result<Vec<Model>, String> {
    select_task_core(env, query).await
}

#[tauri::command]
pub async fn select_tasks(env: String, query: Option<TaskRequest>) -> Result<Vec<Model>, String> {
    select_tasks_core(env, query).await
}

#[tauri::command]
pub async fn update_task(env: String, task: TaskRequest) -> Result<Model, String> {
    update_task_core(env, task).await
}

// cache
#[tauri::command]
pub async fn select_active_tasks(env: String) -> Result<Option<CacheTasks>, String> {
    select_active_tasks_core(env).await
}

#[tauri::command]
pub async fn start_active_tasks(env: String) -> Result<(), String> {
    start_active_tasks_core(env).await
}

#[tauri::command]
pub async fn stop_active_tasks(env: String) -> Result<(), String> {
    stop_active_tasks_core(env).await
}
